src/logger.ts

```diff
@@ -0,0 +1,427 @@
+		// chunks stored in the writable queue of @_gzip and have them in
+		// @_stream instead. This is desirable because once @_gzip.flush() is
+		// called it will do all pending writes with that flush flag. This is
+		// not what we want though. If there are 100 items in the queue we want
+		// to write all of them with Z_NO_FLUSH and only afterwards do a
+		// Z_SYNC_FLUSH to maximize compression
+		this.gzip = zlib.createGzip({ writableHighWaterMark: 1024 } as any);
+		this.gzip.on('error', () => this.teardown());
+		this.gzip.pipe(this.req);
+
+		// Only start piping if there has been no error after the header flush.
+		// Doing it immediately would potentialy lose logs if it turned out that
+		// the server is unavailalbe because @_req stream would consume our
+		// passthrough buffer
+		this.timeout = setTimeout(() => {
+			if (this.gzip != null) {
+				this.stream.pipe(this.gzip);
+				this.flush();
+			}
+		}, RESPONSE_GRACE_PERIOD);
+
+	}, COOLDOWN_PERIOD);
+
+	private snooze = _.debounce(this.teardown, KEEPALIVE_TIMEOUT);
+
+	// Flushing every ZLIB_TIMEOUT hits a balance between compression and
+	// latency. When ZLIB_TIMEOUT is 0 the compression ratio is around 5x
+	// whereas when ZLIB_TIMEOUT is infinity the compession ratio is around 10x.
+	private flush = _.throttle(() => {
+		if (this.gzip != null) {
+			this.gzip.flush(zlib.Z_SYNC_FLUSH);
+		}
+	}, ZLIB_TIMEOUT, { leading: false });
+
+	private teardown() {
+		if (this.req != null) {
+			clearTimeout(this.timeout);
+			this.req.removeAllListeners();
+			this.req.on('error', _.noop);
+			if (this.gzip != null) {
+				this.stream.unpipe(this.gzip);
+				this.gzip.end();
+			}
+			this.req = null;
+		}
+	}
+
+	private write(message: LogMessage) {
+		this.snooze();
+		if (this.req == null) {
+			this.setup();
+		}
+
+		if (this.writable) {
+			this.writable = this.stream.write(JSON.stringify(message) + '\n');
+			this.flush();
+		} else {
+			this.dropCount += 1;
+		}
+	}
+}
+
+interface LoggerConstructOptions {
+	eventTracker: EventTracker;
+}
+
+type LogEventObject = Dictionary<any> | null;
+
+enum OutputStream {
+	Stdout,
+	Stderr,
+}
+
+export class Logger {
+	private writeLock: (key: string) => Bluebird<() => void> = Bluebird.promisify(
+		new Lock().async.writeLock,
+	);
+	private backend: LogBackend | null = null;
+	private eventTracker: EventTracker;
+	private attached: {
+		[key in OutputStream]: { [containerId: string]: boolean }
+	} = {
+		[OutputStream.Stderr]: { },
+		[OutputStream.Stdout]: { },
+	};
+
+	public constructor({ eventTracker }: LoggerConstructOptions) {
+		this.backend = null;
+		this.eventTracker = eventTracker;
+	}
+
+	public init(
+		apiEndpoint: string,
+		uuid: string,
+		deviceApiKey: string,
+		offlineMode: boolean,
+		enableLogs: boolean,
+	) {
+		this.backend = new ResinLogBackend(apiEndpoint, uuid, deviceApiKey);
+		this.backend.offlineMode = offlineMode;
+		this.backend.publishEnabled = enableLogs;
+	}
+
+	public enable(value: boolean) {
```

Comment:
The `value` here is actually expected to be string or boolean (and I think how we call it in deviceState it's always a string), and that's the reason we use checkTruthy on it.

But since you're changing `init` so that the caller is the one that does checkTruthy, you might want to do the same here, and change https://github.com/resin-io/resin-supervisor/blob/master/src/device-state.coffee#L161 to call checkTruthy instead of this function.

Reply:
Yeah I was thinking that, what I've been doing is tightening the types for when it's called by typescript, but keeping the function validation present, for when these are called by the existing typescript. It does make sense though that we would switch it out in the coffeescript too, will do that now.